src/group_theory/is_free_group.lean

```diff
@@ -1,175 +1,131 @@
 /-
 Copyright (c) 2021 David Wärn. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: David Wärn, Eric Wieser
+Authors: David Wärn, Eric Wieser, Joachim Breitner
 -/
 import group_theory.free_group
 /-!
 # Free groups structures on arbitrary types
 
-This file defines the universal property of free groups, and proves some things about
-groups with this property. For an explicit construction of free groups, see
-`group_theory/free_group`.
+This file defines a type class for type that are free groups, together with the usual operations.
+The type class can be instantiated by providing an isomorphim to the canonical free group, or by
+proving that the universal property holds.
+
+For the explicit construction of free groups, see `group_theory/free_group`.
 
 ## Main definitions
 
 * `is_free_group G` - a typeclass to indicate that `G` is free over some generators
-* `is_free_group.lift` - the (noncomputable) universal property of the free group
-* `is_free_group.to_free_group` - any free group with generators `A` is equivalent to
-  `free_group A`.
+* `is_free_group.of` - the canonical injection of `G`'s generators into `G`
+* `is_free_group.lift` - the universal property of the free group
 
-## Implementation notes
+## Main results
 
-While the typeclass only requires the universal property hold within a single universe `u`, our
-explicit construction of `free_group` allows this to be extended universe polymorphically. The
-primed definition names in this file refer to the non-polymorphic versions.
+* `is_free_group.to_free_group` - any free group with generators `A` is equivalent to
+  `free_group A`.
+* `is_free_group.unique_lift` - the universal property of a free group
+* `is_free_group.of_unique_lift` - constructing `is_free_group` from the universal property
 
 -/
-noncomputable theory
-universes u w
-
-/-- `is_free_group G` means that `G` has the universal property of a free group,
-That is, it has a family `generators G` of elements, such that a group homomorphism
-`G →* X` is uniquely determined by a function `generators G → X`. -/
-class is_free_group (G : Type u) [group G] : Type (u+1) :=
-(generators : Type u)
-(of : generators → G)
-(unique_lift' : ∀ {X : Type u} [group X] (f : generators → X),
-                ∃! F : G →* X, ∀ a, F (of a) = f a)
-
-instance free_group_is_free_group {A} : is_free_group (free_group A) :=
-{ generators := A,
-  of := free_group.of,
-  unique_lift' := begin
-    introsI X _ f,
-    have := free_group.lift.symm.bijective.exists_unique f,
-    simp_rw function.funext_iff at this,
-    exact this,
-  end }
 
-namespace is_free_group
+universes u
 
-variables {G H : Type u} {X : Type w} [group G] [group H] [group X] [is_free_group G]
+/-- `is_free_group G` means that `G` isomorphic to a free group. -/
+class is_free_group (G : Type u) [group G] :=
+  ( generators : Type u )
+  ( mul_equiv [] : free_group generators ≃* G )
```

Comment:
It looks like you reverted all the formatting fixes that were suggested in previous review comment

Reply:
Oh, darn how did that happen. Maybe `vscode` didn’t reload the file… I’ll try to restore them.